pkg/audit/manager.go

```diff
@@ -126,47 +126,95 @@ type PubsubMsg struct {
 	ResourceLabels        map[string]string `json:"resourceLabels,omitempty"`
 }
 
-// updateListEntry holds the information necessary to update the
-// audit results in the `status` field of the constraint template.
-// Adding data to this struct has a large impact on memory usage.
-type updateListEntry struct {
-	group             string
-	version           string
-	kind              string
-	namespace         string
-	name              string
-	msg               string
-	enforcementAction util.EnforcementAction
-}
+// A max PriorityQueue implements heap.Interface and holds StatusViolation.
+type SVQueue []*StatusViolation
+
+func (svq SVQueue) Len() int { return len(svq) }
 
-// ByGVKNNMsg implements sort.Interface based on the group, version, kind, name, namespace, and msg fields.
-type byGVKNNMsg []updateListEntry
+// Implements sort.Interface based on the group, version, kind, namespace, name, message and enforcement action fields.
+// For Pop to give us the highest priority, use greater than here.
+func (svq SVQueue) Less(i, j int) bool {
+	if svq[i].Group != svq[j].Group {
+		return svq[i].Group > svq[j].Group
+	}
+	if svq[i].Version != svq[j].Version {
+		return svq[i].Version > svq[j].Version
+	}
+	if svq[i].Kind != svq[j].Kind {
+		return svq[i].Kind > svq[j].Kind
+	}
+	if svq[i].Namespace != svq[j].Namespace {
+		return svq[i].Namespace > svq[j].Namespace
+	}
+	if svq[i].Name != svq[j].Name {
+		return svq[i].Name > svq[j].Name
+	}
+	if svq[i].Message != svq[j].Message {
+		return svq[i].Message > svq[j].Message
+	}
+	return svq[i].EnforcementAction > svq[j].EnforcementAction
+}
 
-func (a byGVKNNMsg) Len() int {
-	return len(a)
+func (svq SVQueue) Swap(i, j int) {
+	svq[i], svq[j] = svq[j], svq[i]
 }
 
-func (a byGVKNNMsg) Less(i, j int) bool {
-	if a[i].group != a[j].group {
-		return a[i].group < a[j].group
-	}
-	if a[i].version != a[j].version {
-		return a[i].version < a[j].version
+func (svq *SVQueue) Push(x any) {
+	sv, ok := x.(*StatusViolation)
+	if !ok {
+		return
 	}
-	if a[i].kind != a[j].kind {
-		return a[i].kind < a[j].kind
+	*svq = append(*svq, sv)
+}
+
+func (svq *SVQueue) Pop() any {
+	old := *svq
+	n := len(old)
+	sv := old[n-1]
+	old[n-1] = nil
+	*svq = old[:n-1]
+	return sv
+}
+
+// LimitQueue implements logic to ensure priority queue len <= limit in order to provide performance guarantees on heap methods.
+type LimitQueue struct {
+	limit int
+	svq   SVQueue
+}
+
+func newLimitQueue(l int) *LimitQueue {
+	lq := LimitQueue{
+		limit: l,
+		svq:   make(SVQueue, 0, l),
 	}
-	if a[i].namespace != a[j].namespace {
-		return a[i].namespace < a[j].namespace
+	heap.Init(&lq.svq)
+	return &lq
+}
+
+func (lq *LimitQueue) Len() int { return lq.svq.Len() }
+
+func (lq *LimitQueue) Push(x *StatusViolation) {
+	heap.Push(&lq.svq, x)
+	for lq.svq.Len() > lq.limit {
+		heap.Pop(&lq.svq)
```

Comment:
Can we avoid having to push then pop here? can we only push when len is < limit?

Reply:
I am not sure, but I think the push may be needed to make sure we get consistent results from sort. I think not pushing when the queue is full may result in not including results that otherwise might have been included in the sorting order.

Reply:
This Push both adds the object to the heap using the svq Push function and performs a sort operation on the heap using the svq Less function. The sort operation ensures that whatever we then Pop will be the max StatusViolation by GVKNNMsg out of the 20 under consideration. As Jaydip mentioned, not pushing when the queue is full means the results will give StatusViolations in sorted order by GVKNNMsg.